File: mimic/src/coords.f90

subroutine distance_1d_float(rx, boxsize, newrx)

  ! Distance in 1D.
  !
  ! Parameters
  ! ----------
  ! rx : float
  !   Radial distance.
  ! boxsize : float
  !   Boxsize.
  !
  ! Returns
  ! -------
  ! newrx : float
  !   Distance with periodic conditions.

  implicit none
  integer, parameter :: dp = kind(1.d0)

  real(kind=dp), intent(in) :: rx, boxsize
  real(kind=dp), intent(out) :: newrx

  if (rx .lt. -boxsize/2.) then
    newrx = rx + boxsize
  else if (rx .gt. boxsize/2.) then
    newrx = rx - boxsize
  else
    newrx = rx
  end if
  newrx = rx

end subroutine distance_1d_float


subroutine distance_3d_float(rx, ry, rz, boxsize, newr, newrx, newry, newrz)

  ! Distance in 3D.
  !
  ! Parameters
  ! ----------
  ! rx, ry, rz : float
  !   Radial distance in each axis.
  ! boxsize : float
  !   Boxsize.
  !
  ! Returns
  ! -------
  ! newr : float
  !   Distance with periodic conditions.
  ! newrx, newry, newrz : float
  !   Distance with periodic conditions in each axes.

  implicit none
  integer, parameter :: dp = kind(1.d0)

  real(kind=dp), intent(in) :: rx, ry, rz, boxsize
  real(kind=dp), intent(out) :: newr, newrx, newry, newrz

  call distance_1d_float(rx, boxsize, newrx)
  call distance_1d_float(ry, boxsize, newry)
  call distance_1d_float(rz, boxsize, newrz)

  newr = sqrt(newrx**2 + newry**2 + newrz**2)

end subroutine distance_3d_float


subroutine get_vec_norm_float(x, y, z, nx, ny, nz)

  ! Normalise vector.
  !
  ! Parameters
  ! ----------
  ! x, y, z : float
  !   Vector.
  !
  ! Returns
  ! -------
  ! nx, ny, nz : float
  !   Normalised vector.

  implicit none
  integer, parameter :: dp = kind(1.d0)

  real(kind=dp), intent(in) :: x, y, z
  real(kind=dp), intent(out) :: nx, ny, nz

  real(kind=dp) :: r

  r = sqrt(x**2 + y**2 + z**2)

  if (r .eq. 0) then
    nx = 1./sqrt(3.)
    ny = 1./sqrt(3.)
    nz = 1./sqrt(3.)
  else
    nx = x/r
    ny = y/r
    nz = z/r
  end if

end subroutine get_vec_norm_float

File: mimic/src/src/coords.py
import numpy as np
from numba import njit


@njit
def distance_1d_float(rx, boxsize):
    """Wrap a 1D separation into the periodic box.

    Parameters
    ----------
    rx : float
        Separation along one axis.
    boxsize : float
        Size of the periodic box.

    Returns
    -------
    float
        Wrapped separation in the range [-boxsize/2, boxsize/2].
    """
    if rx < -boxsize / 2.0:
        return rx + boxsize
    elif rx > boxsize / 2.0:
        return rx - boxsize
    else:
        return rx


@njit
def distance_3d_float(rx, ry, rz, boxsize):
    """Compute a periodic 3D separation and its magnitude.

    Parameters
    ----------
    rx, ry, rz : float
        Coordinate separations along each axis.
    boxsize : float
        Size of the periodic box.

    Returns
    -------
    tuple
        (r, newrx, newry, newrz) where r is the periodic distance and
        newrx/newry/newrz are the wrapped axis separations.
    """
    newrx = distance_1d_float(rx, boxsize)
    newry = distance_1d_float(ry, boxsize)
    newrz = distance_1d_float(rz, boxsize)
    newr = np.sqrt(newrx * newrx + newry * newry + newrz * newrz)
    return newr, newrx, newry, newrz


@njit
def get_vec_norm_float(x, y, z):
    """Normalize a 3D vector, with a fallback for zero length.

    Parameters
    ----------
    x, y, z : float
        Vector components.

    Returns
    -------
    tuple
        Normalized vector components (nx, ny, nz).
    """
    r = np.sqrt(x * x + y * y + z * z)
    if r == 0.0:
        c = 1.0 / np.sqrt(3.0)
        return c, c, c
    return x / r, y / r, z / r
